src/framework/midi/internal/platform/osx/coremidiinport.cpp

```diff
@@ -191,16 +191,16 @@ void CoreMidiInPort::initCore()
         MIDIReceiveBlock receiveBlock = ^ (const MIDIEventList* eventList, void* /*srcConnRefCon*/) {
             const MIDIEventPacket* packet = eventList->packet;
             for (UInt32 index = 0; index < eventList->numPackets; index++) {
+                LOGD() << "midi message size " << packet->wordCount << " bytes";
                 // Handle packet
-                if (packet->wordCount != 0 && packet->wordCount <= 4) {
-                    Event e = Event::fromRawData(packet->words, packet->wordCount);
+                uint32_t cnt = 0;
+                while(cnt < packet->wordCount) {
+                    Event e = Event::fromRawData(&packet->words[cnt], 2);
```

Comment:
In general, I think your change makes sense, but I'm not so sure if you can assume that each group of two `uint32_t`s corresponds to one event. Different types of MIDI messages have different lengths; and I believe (at least MIDI 1.0 messages) the length goes per single byte, not per `uint32_t` (i.e. 4 bytes). 
So parsing a stream of MIDI events is not entirely trivial, and I think that's also the reason that I didn't implement that properly back then (because I hadn't seen any case where a MIDIPacket contained multiple message in practice).

Somehow I'm having trouble finding a good tutorial about this, but Bing Chat generated this:
<details><summary>Click to expand</summary>

```c++
        const Byte *message = &packet->data[0];
        int bytesRemaining = packet->length;
        while (bytesRemaining > 0) {
            int messageLength = 0;
            if ((message[0] & 0x80) == 0x80) {
                // Status byte
                if ((message[0] & 0xF0) == 0xF0) {
                    // System message
                    switch (message[0]) {
                        case 0xF0:
                            // Start of system exclusive message
                            messageLength = 1;
                            while (message[messageLength] != 0xF7) {
                                ++messageLength;
                            }
                            ++messageLength;
                            break;
                        case 0xF1:
                        case 0xF3:
                            // One-byte system common message
                            messageLength = 2;
                            break;
                        case 0xF2:
                            // Two-byte system common message
                            messageLength = 3;
                            break;
                    }
                } else {
                    // Channel message
                    switch (message[0] & 0xF0) {
                        case 0x80:
                        case 0x90:
                        case 0xA0:
                        case 0xB0:
                        case 0xE0:
                            // Three-byte channel message
                            messageLength = 3;
                            break;
                        case 0xC0:
                        case 0xD0:
                            // Two-byte channel message
                            messageLength = 2;
                            break;
                    }
                }
            } else {
                // Data byte
                messageLength = 1;
            }
            if (messageLength == 0) {
                // Invalid message
                break;
            }
            // Process the MIDI message here
            printf("MIDI message: ");
            for (int j = 0; j < messageLength; ++j) {
                printf("%02X ", message[j]);
            }
            printf("\n");
            message += messageLength;
            bytesRemaining -= messageLength;
        }
```

</details>

I also found this: https://learn.sparkfun.com/tutorials/midi-tutorial/all#implementing-midi

Reply:
Thank you for your fast reply! You're right we should probably implement that correctly. Since in this case we're [requesting MIDI 2.0 from macOS here](https://github.com/musescore/MuseScore/blob/546bbf6b1184c4f75ce8343c79e92526a7ce14c3/src/framework/midi/internal/platform/osx/coremidiinport.cpp#L209) it does all the conversion and we're getting MIDI 2.0 messages.

I don't really know much about MIDI 2.0 but the nice thing is that the spec explicitly defines the size of each message by the value of the first 4 bit according to this table:

<img width="660" alt="Screenshot 2023-12-20 at 10 23 52" src="https://github.com/musescore/MuseScore/assets/3973466/fe31de97-ba21-4ab1-a766-2dd191242fe6">

So I think this should be relatively easy. I think the code you genereted is really just treating MIDI 1.0 and it's system exclusives which is more complicated. I'm not entirely sure how system exclusives map to MIDI 2.0 but I think it is also not our most important concern.

I try to implement it properly later today or tomorrow.